utils/controller_debug.py

```diff
@@ -0,0 +1,1058 @@
+            while len(SendCmd(OP_PBREAD)) > 0:
+                pass
+            cl.remove("--flush")
+
+        if len(cl) > 0:
+            fp = open(cl[0], "w")
+        else:
+            fp = None
+
+        try:
+            while True:
+                dat = SendCmd(OP_PBREAD)
+                if len(dat) < 1:
+                    continue
+
+                S = "".join([chr(x) for x in dat])
+
+                sys.stdout.write(S)
+                sys.stdout.flush()
+
+                if fp != None:
+                    fp.write(S)
+
+        except KeyboardInterrupt:
+            print()
+            ReSync()
+            pass
+
+    def do_EOF(self, line):
+        return True
+
+    def do_exit(self, line):
+        return True
+
+    def do_get(self, line):
+        cl = line.split()
+        if len(cl) < 1:
+            print("Please give the variable name to read")
+            return
+
+        print(GetVar(cl[0]))
+
+    def complete_get(self, text, line, begidx, endidx):
+        var = text
+
+        out = []
+        for i in varDict.keys():
+            if i.startswith(var):
+                out.append(i)
+
+        return out
+
+    def help_get(self):
+        global varDict
+        print("Read the value of a debug variable and display it\n")
+        print("Variables currently defined:")
+        for k in varDict.keys():
+            print("   %-10s - %s" % (k, varDict[k].help))
+
+    def do_set(self, line):
+        cl = line.split()
+        if len(cl) < 2:
+            print("Please give the variable name and value")
+            return
+        SetVar(cl[0], cl[1])
+
+    def complete_set(self, text, line, begidx, endidx):
+        var = text
+
+        out = []
+        for i in varDict.keys():
+            if i.startswith(var):
+                out.append(i)
+
+        return out
+
+    def help_set(self):
+        global varDict
+        print("Set the value of a debug variable\n")
+        print("Variables currently defined:")
+        for k in varDict.keys():
+            print("   %-10s - %s" % (k, varDict[k].help))
+
+    def do_trace(self, line):
+        """
+       This command is used to access the trace buffer on the controller.
+       A sub-command must be passed as an option:
+
+       trace flush
+         Stops the trace if one was on-going and flushes the trace buffer
+
+       trace start
+         Starts the trace collecting data.
+
+       trace graph
+         This will download the data and display it graphically
+
+       trace download <filename>
+         This will download the data and save it to a file with the given
+         name.  If no file name is given, then trace.dat will be used
+
+       The trace feature allows variables to be sampled in real time and saved
+       to a large internal memory buffer in the drive.  This command can then
+       be used to download the captured data and save or display it.
+
+       The trace first must be set up by setting specific variables.
```

Comment:
How does one set these variables?

Reply:
Oh, upon looking at the C++ I'm realizing that this talks about the controller's variables, i.e. set using the regular "set" command. I thought this means set some variables in the Python program. Worth clarifying